### ISSUE TITLE
C.148 only solves issues stemming from bad design

### ISSUE THREAD

**Issue by scraimer** [isocpp/CppCoreGuidelines#910]
> C.148: Use dynamic_cast to a pointer type when failure to find the required class is considered a valid alternative

Full disclosure: I do this myself in production code all the time.

That being said, it's not ideal. I only ever have to do it because I'm dealing with legacy design.

While trying to come up with an example for C.148, every example I thought up indicated a bad design decision.

### Example 1: Containers of heterogeneous types

In a game world, there can be many "entities": monsters, walls, health potions, whatever. It makes sense to put them all in a single container to make it easy to iterate over all of them. Some entities allow the player to walk through them. Some entities deal damage to a player that touches them. As long as the common interface allows to apply the common functions, everything is great.

We'd apply C.148 in this case to check whether the entity implements additional functionality. For example, using `dynamic_cast` like so:

    for( entity * ent : world )
    {
       if( player.is_touching( ent ) )
       {
          if( dynamic_cast<health_potion *>( ent ) )
          {
             player.increase_health();
          }
          else if( dynamic_cast<monster *>( ent ) )
          {
             player.decrease_health();
          }
       }
    }

Doesn't this indicate an issue with the design? It would have been far better to have a virtual function called `get_effect`, and have it overridden in each subclass of `entity`.

    for( entity * ent : world )
    {
       if( player.is_touching( ent ) )
       {
           auto effect = ent->get_effect();
           player.apply_effect( effect );
       }
    }

### Example 2: Arguments in function calls

Here's a function for a game, perhaps an earlier version of the game in Example 1. Notice how the `dynamic_cast` is used to determine whether to apply extra effects:

    void player::apply_effect( entity const * const ent )
    {
        // ... do some common effects
    
        if( auto hb = dynamic_cast< health_box const * const >( ent ); hb )
        {
           increase_health();
        }

        if( auto m = dynamic_cast< monster const * const >( ent ); m )
        {
           decrease_health();
        }
    }

In essence, we have rolled several different functions into a single function. Depending on the arguments, the `apply_effect` might do wildly different things.

* Instead of having one function that does multiple different things, left have separate functions: use overloads.
* Better, create a virtual function in `entity` that is overridden when needed.
* Better yet, If the type is known at compile time at the call site, then we shouldn't even need the `dynamic_cast`, and we can solve it using a set of function template specializations.

----

In summary, while I agree with C.148, I cannot think of a good design in modern C++ that would require it. As far as I can tell, it's a great solution for dealing with issues in an existing design. But as a guideline for "what would you like your code to look like [...] given that you can start now", it doesn't seem right. 

C.148 should either be removed or re-worded so it doesn't conflict with the intent of the document as stated in the section "Abstract".

Please, tell me I'm wrong!

**Comment by brenoguim** [comment]
Using the good old Shape as base class, Triangle and Circle deriving from it, wouldn't the "virtual equal(const Shape&) const = 0;" function implementation require dynamic_cast?

**Comment by galik** [comment]
If you are having to `dynamic_cast` a pointer in order to figure out what type it is so that you can call the correct functions, then that's a bit of a *polymorphic* failure. I dare say *sometimes* it's going to happen in those parts of a system where *polymorphism* isn't the right tool for the job because the actual types need to be known.

One way to deal with that is to have different *views* of an object. You can have a master *vector* of *unique pointers* managing the life times of all your *polymorphic* types. In addition to that, for other parts of the system that need to know actual types, you can have properly typed *vectors* of *raw pointers* to the same objects.

Then those parts of the system that work well *polymorphically* can use the master vector whilst those areas of the system that need to know the actual type can use the correctly typed *view vectors*.

There is an example here:
https://stackoverflow.com/questions/32022884/code-refactoring-with-existing-classes-in-c/32023858#32023858
(warning, example doesn't follow these guidelines very well)

 In that example `dynamic_cast` is used to create the strongly typed *view vectors*.

**Comment by scraimer** [comment]
@brenoguim 
I was talking about C.148 (`dynamic_cast` of pointers) not C.147 (`dynamic_cast` of references)

**Comment by brenoguim** [comment]
@scraimer , I'm not sure my example falls exactly on the C.147, even though I used a reference on the parameter.

C.147 says "when failure to find the required class is considered an error", which is not the case for the "equal" method.
The implementation of "equal" method would most likely use a dynamic_cast of pointer, and if the result is nullptr, just return false. So getting a different type is a perfectly valid alternative, best fitting into C.148 case.

**Comment by scraimer** [comment]
@galik 
Good example! I have a nearly identical example in my project's production code. It's one of the bad designs choices I was talking about. In my case, allocating everything into a single "owning" vector.

In the example you gave, there's the function `add_item` which uses `dynamic_cast` create observer pointers, sorted into multiple baskets, what you called "view vectors".

But the call to `add_item` is done in a site where the type of the argument is known at compile-time. So there's no need for a `dynamic_cast`. Instead, it can be solved with a template specialization and a `static_cast`. (I'm of course, assuming the example is modernized into modern C++, hopefully to remove the use of `new`, which would have shifted the pointers to become references, too.)

**Comment by gdr-at-ms** [comment]
@scraimer Could you expand on why you consider the template specializations coupled with `static_cast` a better or superior design?  I am curious because `static_cast` isn't a completely safe operation whereas `dynamic_cast` does not let you a type confusion pass by.

**Comment by scraimer** [comment]
@brenoguim 
I'd much rather design the classes differently. Any place where you would put `dynamic_cast` inside `equals` and it would fail, I'd much rather have a compilation error.

Every call to `equals` really has three results: true, false, and undefined. I'm saying this because I don't see any case where a Circle and a Triangle are truly equal. If you just want to compare common attributes of theirs, then that attribute should belong in the common ancestor, Shape.

**Comment by brenoguim** [comment]
@scraimer , I'm not sure if I made the example clear enough:
```
struct Shape {
    virtual ~Shape() {}
    virtual bool equal(const Shape&) const = 0;
};

struct Square : public Shape {
    bool equal(const Shape& s) const override {
        if (const Square* sq = dynamic_cast<const Square*>(&s)) {
            return m_side == sq->m_side;
        } else {
            return false;
        }
    }
    unsigned m_side;
};
```
Calls for "equals" only have two results: true or false. Comparing the Square to any other shape is very much defined, as they are different.

In all honesty, I'm not sure if this the best example to show usage of dynamic_cast. I just picked the first thing I remember about the production code I work with and that I don't know how to change to avoid the cast.

**Comment by scraimer** [comment]
@gdr-at-ms 
In the [example that @galik showed](https://stackoverflow.com/questions/32022884/code-refactoring-with-existing-classes-in-c/32023858#32023858), the scenario presented happens during the initialization of the application. I thought to take advantage of having all the type information at the call site during compile-time. But... I failed to pay attention! The multiple inheritance makes `static_cast` is unsafe, and my solution is not good.

After thinking about it for a bit, all my alternative solutions to this example are isomorphic to doing the run-time check of `dynamic_cast`, as long as `add_items` must accept objects via `Role *`, and not as the original object's type.

**Comment by scraimer** [comment]
OK, so I'll use a similar example for C.148 and construct a PR. Thanks!

**Comment by brenoguim** [comment]
This example looks more complicated than necessary to demonstrate the point raised in C.148...
